espnet/nets/beam_search_timesync_streaming.py

```diff
@@ -0,0 +1,337 @@
+        self.attn_cache = dict()
+        self.lm_cache = dict()
+        self.enc_output = enc_output
+        self.sos_th = self.sos_th.to(enc_output.device)
+
+        if self.decoder is not None:
+            init_decoder_state = self.decoder.init_state(enc_output)
+            decoder_scores, decoder_state = self.decoder.score(
+                self.sos_th, init_decoder_state, enc_output
+            )
+            self.attn_cache[(self.sos,)] = CacheItem(
+                state=decoder_state,
+                scores=decoder_scores,
+                log_sum=0.0,
+            )
+
+            # hacked
+            decoder_scores, decoder_state = self.decoder.score(
+                torch.tensor([self.sos, 250003], device=enc_output.device),
+                init_decoder_state,
+                enc_output,
+            )
+            self.attn_cache[(self.sos, 250003)] = CacheItem(
+                state=decoder_state,
+                scores=decoder_scores,
+                log_sum=0.0,
+            )
+        if self.lm is not None:
+            init_lm_state = self.lm.init_state(enc_output)
+            lm_scores, lm_state = self.lm.score(self.sos_th, init_lm_state, enc_output)
+            self.lm_cache[(self.sos,)] = CacheItem(
+                state=lm_state,
+                scores=lm_scores,
+                log_sum=0.0,
+            )
+
+    def cached_score(
+        self,
+        h: Tuple[int],
+        cache: dict,
+        scorer: ScorerInterface,
+        recompute_cache: bool = False,
+    ) -> Any:
+        """Retrieve decoder/LM scores which may be cached."""
+        root = h[:-1]  # prefix
+
+        # if (root in cache and not recompute_cache) or len(root) == 1:
+        #     if root not in self.block_set:
+        #         import pdb;pdb.set_trace()
+        if (root in cache and root in self.block_set) or len(root) <= 1:  # hacked 1-->2
+            logging.debug("not recomputing")
+            root_scores = cache[root].scores
+            root_state = cache[root].state
+            root_log_sum = cache[root].log_sum
+        else:  # run decoder fwd one step and update cache
+            root_root = root[:-1]
+            root_root_state = cache[root_root].state
+            root_scores, root_state = scorer.score(
+                torch.tensor(root, device=self.enc_output.device).long(),
+                root_root_state,
+                self.enc_output,
+            )
+            root_log_sum = cache[root_root].log_sum + float(
+                cache[root_root].scores[root[-1]]
+            )
+            cache[root] = CacheItem(
+                state=root_state, scores=root_scores, log_sum=root_log_sum
+            )
+        cand_score = float(root_scores[h[-1]])
+        score = root_log_sum + cand_score
+
+        logging.debug("cand score: " + str(cand_score))
+        logging.debug("decoder score: " + str(score))
+
+        self.block_set.add(root)
+
+        return score
+
+    def joint_score(
```

Comment:
Maybe better to make this part generalized (can join multiple scores with some weights like label-sync beam search) in the future.
You can leave some comments (TODO?)